### DISCUSSION TITLE
Trim image to smallest square (centered)

### DISCUSSION THREAD

**Discussion by SheriffBuzz** [ImageMagick/ImageMagick#5877] [Help] [answered]
I am using version 7.1.0-51 Q16-HDRi x64 (windows)

I am trying to trim an image to a smallest square (centered), removing as much transparent background as possible.

[SourceImageWithTransparency](https://user-images.githubusercontent.com/83767022/207117828-d0589148-70f2-4649-b3e1-fe15da410ce7.png)

My input image is always a square. Like this:
<img src="https://user-images.githubusercontent.com/83767022/207106578-f4f3d407-3cc6-4960-bb16-5168c0be333a.png" width=500 height=500>

I was using [this solution](https://stackoverflow.com/a/49100330/13273420) but it doesnt seem to do what I think it does, it isnt centered around the original center point.

convert input.png -trim -background none \
   -set page "%[fx:max(w,h)]x%[fx:max(w,h)]+%[fx:(max(w,h)-w)/2]+%[fx:(max(w,h)-h)/2]" \
   -coalesce output.png

Im not well versed on the different versions of ImageMagick but the actual command I was running is the same as above with "magick " before it. Is it possible the command I am using was for a different version?

This gives me:
<img src="https://user-images.githubusercontent.com/83767022/207106757-efb25017-4fd9-4fbe-b452-f028682dc374.png" width=500 height=500>

Is there a way to get this?
<img src="https://user-images.githubusercontent.com/83767022/207106992-c59c135e-dcbf-46ed-886c-fdc46a9fdd53.png" width=500 height=500>

**Comment by fmw42** [comment]
Post your actual transparent image.  What you have posted has the transparency replaced with checkerboard.

**Reply by SheriffBuzz** [reply]
I have added it now at the top of the post

**Comment by fmw42** [comment, marked as answer]
I think this does what you want.

```
Compute the center (cx,cy) and the trimbox.
Compute the maximum distance from the center to the sides of the trimbox.
Compute the top left and top right of the crop box
Use twice the maximum distance as the dropbox width and height.

```

Unix (Linux, Mac OSX) Syntax:

```
infile="207117828-d0589148-70f2-4649-b3e1-fe15da410ce7.png"
declare `convert "$infile" +repage \
-format "cx=%[fx:round(w/2)]\ncy=%[fx:round(h/2)]\ntrimbox=%@\n" +repage info:`
twd=`echo $trimbox | tr "x" "+" | cut -d+ -f1`
tht=`echo $trimbox | tr "x" "+" | cut -d+ -f2`
txo=`echo $trimbox | tr "x" "+" | cut -d+ -f3`
tyo=`echo $trimbox | tr "x" "+" | cut -d+ -f4`
upper=$((cy-tyo))
lower=$((tyo+tht-1-cy))
left=$((cx-txo))
right=$((txo+twd-1-cx))
dist=`convert xc: -format "%[fx:round(max($upper,max($lower,max($left,$right))))]" info:`
xo=$(($cx-$dist))
yo=$(($cy-$dist))
dist=$((2*dist))
cropvals="${dist}x${dist}+${xo}+${yo}"
convert "$infile" +repage -crop $cropvals +repage result.png

```

![result](https://user-images.githubusercontent.com/13458218/207125102-c79947c8-2e30-4641-8067-bf53c31f337c.png)

**Reply by SheriffBuzz** [reply]
I'm on Windows 10 and this script didn't seem to be working when running from Git Bash, not really sure how I was supposed to go about running it, if at all on Windows.

However, I converted all the logic into a Powershell script and it works great!

**Reply by fmw42** [reply]
As I said it was a Unix script.  It does not run on Windows.  Windows scripting is different.Sorry, but I do not use Windows and do not know how to script it.  Good that you were able to convert it yourself.
